subsys/bluetooth/host/l2cap.c

```diff
@@ -1032,14 +1033,42 @@ static uint16_t l2cap_chan_accept(struct bt_conn *conn,
 	return BT_L2CAP_LE_SUCCESS;
 }
 
-static bool l2cap_check_security(struct bt_conn *conn,
+static uint16_t l2cap_check_security(struct bt_conn *conn,
 				 struct bt_l2cap_server *server)
 {
 	if (IS_ENABLED(CONFIG_BT_CONN_DISABLE_SECURITY)) {
-		return true;
+		return BT_L2CAP_LE_SUCCESS;
 	}
 
-	return conn->sec_level >= server->sec_level;
+	if (conn->sec_level >= server->sec_level) {
+		return BT_L2CAP_LE_SUCCESS;
+	}
+
+	if (conn->sec_level > BT_SECURITY_L1) {
+		return BT_L2CAP_LE_ERR_AUTHENTICATION;
+	}
```

Comment:
I am not sure I follow this statement. 

If e.g. `conn->sec_level` is `BT_SECURITY_L3` and `server->sec_level` is BT_SECURITY_L4 this will return with `BT_L2CAP_LE_ERR_AUTHENTICATION`, but couldn't it just as likely fail with `BT_L2CAP_LE_ERR_KEY_SIZE`?

Reply:
If I get the code correctly, key size is verified by server user in accept callback

Reply:
But then that will leave this function to assume that a specific security check is done elsewhere. It's just a matter of responding with the right error code, so it's not a huge deal, but I do think that this check could possibly (should?) be expanded to check the key in the specific case that if current security is L3 and required security is L4, that we check if the the key size is correct or not.

Reply:
if you are referring to L4 requiring 16 key size than this is check by update_sec_level(),  ie connection security will not be set to L4 if keysize is not max

Reply:
I am referring to that, but in the case where the ACL connection is only L3 and the L2CAP connection requires L4.

Reply:
there is no explicit test for that and spec doesn't explicitly mention that, but since it is security level that was not reached I'd keep this as is (insufficient authentication) - it is just co-incidental that L4 == (L3 + SC + keysize 16)

My understanding of 'insufficient key size'  error is that security level is reached, but only key size is too small.

Reply:
@sjanc The quote from the spec you use in your commit message actually contains this explicitly: 

> If an LTK or an STK is available and encryption is required (LE security mode 1) but encryption is not enabled, the service request shall be rejected with the error code “Insufficient Encryption”. **If the encryption is enabled with a key size that is too short then the service request shall be rejected with the error code “Encryption Key Size Too Short.”**

Reply:
L3 can also be legacy pairing, in which case we must return 'Insufficient Authentication'

Reply:
I don't think that's relevant to this discussion about Level 4, which is Secure Connections with 128-bit keys, where it can "fail" on either the key size, authentication or encryption.

Reply:
Moving check from application provided accept context into this check could have potentially affect application behavior (and thus API). This is going to be backported into LTS branch and aims at fixing qualification, not reworking L2CAP API.

Reply:
> Moving check from application provided accept context into this check could have potentially affect application behavior (and thus API). This is going to be backported into LTS branch and aims at fixing qualification, not reworking L2CAP API.

I'm worried we may be talking past each other. If the application supplies `BT_SECURITY_L4` as the required security, why shouldn't that be checked in the same way that `BT_SECURITY_L3` is?

Reply:
that said, I'm not sure why it was designed that way, and I'm fine with moving check into stack,  just in separate PR

Reply:
> > Moving check from application provided accept context into this check could have potentially affect application behavior (and thus API). This is going to be backported into LTS branch and aims at fixing qualification, not reworking L2CAP API.
> 
> I'm worried we may be talking past each other. If the application supplies `BT_SECURITY_L4` as the required security, why shouldn't that be checked in the same way that `BT_SECURITY_L3` is?

but it is checked the same way (or maybe I'm misunderstanding what is the problem here)

Reply:
Adding that in a separate PR would be fine with me, but could you please add a TODO here so that we don't completely forget it?

Reply:
I've created issue for this  https://github.com/zephyrproject-rtos/zephyr/issues/40216